### PR TITLE
Add "Create from template" option

### PR DESCRIPTION
**Pull request by alexweininger**
* We plan on adding descriptions to the repos on GitHub and then displaying these in the Quickpick. Once these descriptions are added, I'll make a PR for that.
* The angular-basic template no longer has working defaults. This is because with this flow, we cannot pick up the default values from the settings.json in the template since it's not cloned locally.
* @craigshoemaker is a maintainer of the https://github.com/staticwebdev account. I've met with him and they will make sure to let us test any changes/additions they want to make to the templates.
* This is still just an idea, but there might be templates in the future that we don't want to include in our extension. And I think we could use [GitHub repo topics](https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/classifying-your-repository-with-topics) to mark whether to include or exclude a repo from our Quickpick.

### CODE REVIEW

File: src/utils/gitHubUtils.ts [hunk 1]
```diff
@@ -105,3 +105,33 @@ export async function createFork(context: IActionContext, remoteRepo: ReposGetRe
...
+        return client.rest.repos.createUsingTemplate({
+            template_owner: nonNullProp(templateRepo, 'owner').login,
+            template_repo: templateRepo.name,
+            name: newRepoName ?? templateRepo.name,
+            private: isPrivate ?? false,
+            description: description ?? undefined
+        });
+    } catch (e) {
+        const generateUrl: string = `${templateRepo.html_url}/generate`;
+        throw new Error(localize('createFromTemplateFail', 'Could not create repository from template. [Click here to create it manually]({0}).', generateUrl));
```
Comment (neelip) [inline src/utils/gitHubUtils.ts:125]: Maybe this can be "Create it manually" hyperlink instead

File: src/utils/workspaceUtils.ts [hunk 2]
```diff
@@ -66,15 +67,22 @@ export async function tryGetWorkspaceFolder(context: IActionContext): Promise<Wo
...
+    const result = await context.ui.showWarningMessage(message, { modal: true, stepName: noWorkspaceWarning }, openExistingProjectMsg, cloneProjectMsg, createFromTemplateMsg);
     if (result === cloneProjectMsg) {
         await cloneRepo(context, '');
-        context.telemetry.properties.noWorkspaceResult = 'cloneProject';
+        context.telemetry.properties.noWorkspaceResult = cloneProject;
     } else if (result === openExistingProjectMsg) {
         await openFolder(context)
         context.telemetry.properties.noWorkspaceResult = openExistingProject;
+    } else if (result === createFromTemplateMsg) {
+        await createStaticWebAppFromTemplate(context);
```
Comment (nturinski) [inline src/utils/workspaceUtils.ts:84]: Downside of calling `createStaticWebAppFromTemplate` here like this is the verifying workspace notification doesn't go away during the whole create process (if your entry point is the +)

I think just set the `context.isTemplate` to true here and just return undefined and go through the rest of the create process, if we refactor `createStaticWebApp` a bit.  I've left a comment about that.

File: src/commands/createStaticWebApp/createStaticWebAppFromTemplate.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ *--------------------------------------------------------------------------------------------*/
+
+import { IActionContext, ICreateChildImplContext } from 'vscode-azureextensionui';
+import { ext } from '../../extensionVariables';
+import { StaticWebAppTreeItem } from '../../tree/StaticWebAppTreeItem';
+import { SubscriptionTreeItem } from '../../tree/SubscriptionTreeItem';
+import { showSwaCreated } from '../showSwaCreated';
+import { IStaticWebAppWizardContext } from './IStaticWebAppWizardContext';
+
+export async function createStaticWebAppFromTemplate(context: IActionContext & Partial<ICreateChildImplContext> & Partial<IStaticWebAppWizardContext>, node?: SubscriptionTreeItem): Promise<StaticWebAppTreeItem> {
```
Comment (nturinski) [inline src/commands/createStaticWebApp/createStaticWebAppFromTemplate.ts:13]: I think we could refactor this and add it to `createStaticWebApp.ts`.  Kind of similar to how `createStaticWebAppAdvanced`, we could call `createStaticWebApp` and pass in the options to set `context.fromTemplate = true`.

To look cleaner, we could refactor this code:
```
const progressOptions: ProgressOptions = {
        location: ProgressLocation.Notification,
        title: localize('verifyingWorkspace', 'Verifying workspace...')
    };
    await window.withProgress(progressOptions, async () => {
        const folder = await tryGetWorkspaceFolder(context);
        if (folder) {
            await setWorkspaceContexts(context, folder);
        } else {
            await showNoWorkspacePrompt(context);
        }
    });
```
into its own function, and if we see context.fromTemplate === true, we just skip it and go straight down to `node.createChild`.

Another benefit to doing this is we don't have to duplicate the create process at all.  I'm not sure if it was omitted intentionally, but this function is missing the postCreate actions (such as verifying the actions built and opening the node up for the user).  So anytime we wanted to change that behavior, we would have to do it in two places.

File: src/constants.ts [hunk 4]
```diff
@@ -31,6 +33,8 @@ export const isEndGroup = (t: string): boolean => /##\[endgroup\]/.test(t)
 export const githubAuthProviderId: string = 'github';
 export const githubScopes: string[] = ['repo', 'workflow', 'admin:public_key'];
 
+export const templateReposUsername: string = 'staticwebdev';
```
Comment (nturinski) [inline src/constants.ts:36]: I think this is only used in the one spot, so we probably don't need to throw this in the constants file.  We don't want to unnecessarily clutter this file up.

File: src/utils/gitHubUtils.ts [hunk 5]
```diff
@@ -105,3 +105,33 @@ export async function createFork(context: IActionContext, remoteRepo: ReposGetRe
         throw new Error(localize('forkFail', 'Could not automatically fork repository. Please fork [{0}]({1}) manually.', remoteRepo.name, remoteRepo.html_url));
     }
 }
+
+export async function createRepoFromTemplate(context: IActionContext, templateRepo: RepoData, newRepoName?: string, isPrivate?: boolean, description?: string): Promise<ReposCreateUsingTemplateResponse> {
```
Comment (nturinski) [inline src/utils/gitHubUtils.ts:109]: I think we can just include this in the `RepoCreateFromTemplateStep`.  Assuming we refactor it how I said in the comment below, it won't have to be called in two separate places.

File: src/utils/gitHubUtils.ts [hunk 6]
```diff
@@ -105,3 +105,33 @@ export async function createFork(context: IActionContext, remoteRepo: ReposGetRe
...
+        return client.rest.repos.createUsingTemplate({
+            template_owner: nonNullProp(templateRepo, 'owner').login,
+            template_repo: templateRepo.name,
+            name: newRepoName ?? templateRepo.name,
+            private: isPrivate ?? false,
+            description: description ?? undefined
+        });
+    } catch (e) {
+        const generateUrl: string = `${templateRepo.html_url}/generate`;
+        throw new Error(localize('createFromTemplateFail', 'Could not create repository from template. [Create it manually]({0}).', generateUrl));
```
Comment (nturinski) [inline src/utils/gitHubUtils.ts:125]: Have you seen an error thrown from GitHub yet?  Usually GitHub throws pretty reasonable/understandable errors so I don't know if its necessary for us to catch and throw our own.  Plus, they may have more insight as to why it failed.

File: src/utils/gitHubUtils.ts [hunk 7]
```diff
@@ -105,3 +105,33 @@ export async function createFork(context: IActionContext, remoteRepo: ReposGetRe
...
+    }
+}
+
+export async function getTemplateRepos(context: IActionContext): Promise<RepoData[]> {
```
Comment (nturinski) [inline src/utils/gitHubUtils.ts:129]: I think this can also go in the `TemplateListStep` file

File: package.nls.json [hunk 8]
```diff
@@ -1,6 +1,7 @@
 {
     "staticWebApps.createStaticWebApp": "Create Static Web App...",
     "staticWebApps.createStaticWebAppAdvanced": "Create Static Web App... (Advanced)",
+    "staticWebApps.createStaticWebAppFromTemplate": "Create Static Web App from template...",
```
Comment (nturinski) [inline package.nls.json:4]: nit: What do you think about `Create Static Web App... (Template)` or `Create Static Web App from Template...` Somehow the lower case t really bugs me aesthetically lol

File: src/commands/createRepo/RepoCreateFromTemplateStep.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import { Progress } from 'vscode';
+import { AzureWizardExecuteStep } from "vscode-azureextensionui";
+import { ext } from '../../extensionVariables';
+import { RepoData } from '../../gitHubTypings';
+import { createRepoFromTemplate } from "../../utils/gitHubUtils";
+import { localize } from '../../utils/localize';
+import { nonNullProp } from '../../utils/nonNull';
+import { IStaticWebAppWizardContext } from '../createStaticWebApp/IStaticWebAppWizardContext';
+
+export class RepoCreateFromTemplateStep extends AzureWizardExecuteStep<IStaticWebAppWizardContext> {
```
Comment (nturinski) [inline src/commands/createRepo/RepoCreateFromTemplateStep.ts:15]: I'm a little torn here.  On one hand, this step has a few things different happening when compared to RepoCreateStep, but on the otherhand, a lot of it including the set-up and strings are identical.  

I think I'm okay with either solution, but lean a little bit more towards having one step that has a createRepo and createRepoFromTemplate functions that we can call inside the function.

File: src/commands/createStaticWebApp/TemplateListStep.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+import { AzureWizardPromptStep, IAzureQuickPickItem } from 'vscode-azureextensionui';
+import { RepoData } from '../../gitHubTypings';
+import { getTemplateRepos } from '../../utils/gitHubUtils';
+import { localize } from '../../utils/localize';
+import { nonNullProp } from '../../utils/nonNull';
+import { IStaticWebAppWizardContext } from './IStaticWebAppWizardContext';
+
+export class TemplateListStep extends AzureWizardPromptStep<IStaticWebAppWizardContext> {
+    public async prompt(context: IStaticWebAppWizardContext): Promise<void> {
```
Comment (nturinski) [inline src/commands/createStaticWebApp/TemplateListStep.ts:14]: To make it cache template entries, you can do something like this:

```
private _templateRepos: RepoData[] | undefined;
    public async prompt(context: IStaticWebAppWizardContext): Promise<void> {

        const placeHolder: string = localize('chooseTemplatePrompt', 'Choose a template for the new static web app.');

        const getPicks = async () => {
            this._templateRepos ||= await getTemplateRepos(context);
            return this._templateRepos.map((repo: RepoData) => ({ label: repo.name, data: repo }));
        }
```

That way, you wont' have to add more things to the context.

File: package.json [hunk 11]
```diff
@@ -58,6 +59,12 @@
     "main": "./main.js",
     "contributes": {
         "commands": [
+            {
+                "command": "staticWebApps.createStaticWebAppFromTemplate",
+                "title": "%staticWebApps.createStaticWebAppFromTemplate%",
+                "category": "Azure Static Web Apps",
+                "icon": "$(add)"
```
Comment (nturinski) [inline package.json:66]: Does this icon show up anywhere?  I think this is only for the VS Code ribbon.

File: src/utils/gitHubUtils.ts (hunk 6 above)
Comment (alexweininger) [reply src/utils/gitHubUtils.ts:125]: I haven't seen an error. I'll remove that. 

I was sorta following how we did it for `createFork`. Was that for a specific error we were seeing?

https://github.com/microsoft/vscode-azurestaticwebapps/blob/0fb1c6358fea1897be94c246f86e9ceff18961c5/src/utils/gitHubUtils.ts#L105-L106

File: package.nls.json (hunk 8 above)
Comment (alexweininger) [reply package.nls.json]: Changed to `Create Static Web App from Template...`.

File: src/utils/gitHubUtils.ts (hunk 6 above)
Comment (nturinski) [reply src/utils/gitHubUtils.ts]: Yeah, we were running into some issues where repos couldn't be forked.  I don't remember the exact details as to how it was happening, but it was tracked in this issue: https://github.com/microsoft/vscode-azurestaticwebapps/issues/356

File: package.nls.json [hunk 12]
```diff
@@ -1,6 +1,7 @@
 {
     "staticWebApps.createStaticWebApp": "Create Static Web App...",
     "staticWebApps.createStaticWebAppAdvanced": "Create Static Web App... (Advanced)",
+    "staticWebApps.createStaticWebAppFromTemplate": "Create Static Web App from Template...",
```
Comment (alexweininger) [inline package.nls.json:4]: We probably want to change the wording from "template" to "quickstart". But need help deciding.

* "Create Static Web App with a Quickstart"
* "Static Web App Quickstart"
* "Create with Quickstart Template"
* ???

**Pull request closed without merging**